# Managing large and complex training datasets

Large and complex datasets can be difficult to use for model training because of computational constraints; simply loading such datasets may require substantial amounts of RAM.

In addition, the time required to select genes for model training increases as dataset size grows.

However, using more training data does not always lead to a substantial improvement in model performance.

This tutorial presents a method for overcoming these computational limitations when training scAdam models on large, complex datasets with many samples.

In [1]:
# Python packages
import warnings
warnings.simplefilter('ignore')

import scanpy as sc
import scparadise
import numpy as np
import pandas as pd
import os

sc.set_figure_params(dpi = 120)

In [2]:
# Create folder to save files
# single nuclear RNA-seq of human retina
os.makedirs('snRNAseq_human_retina', exist_ok=True)

In [3]:
# Download CELLxGENE dataset (snRNA-seq of human retina): 
# https://cellxgene.cziscience.com/collections/4c6eaf5c-6d57-4c76-b1e9-60df8c655f1e
!wget https://datasets.cellxgene.cziscience.com/2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad

--2025-02-08 13:49:05--  https://datasets.cellxgene.cziscience.com/2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad
Resolving datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)... 52.85.49.24, 52.85.49.28, 52.85.49.17, ...
Connecting to datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)|52.85.49.24|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 37946797973 (35G) [binary/octet-stream]
Saving to: ‘2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad’

2e910e62-7eaf-4c06- 100%[===================>]  35.34G  25.1MB/s    in 27m 20s 

2025-02-08 14:13:58 (22.1 MB/s) - ‘2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad’ saved [37946797973/37946797973]



In [4]:
# Open the dataset in backed mode to inspect the cell-type annotations 
# without loading the full expression matrix into memory.
adata = sc.read_h5ad('2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad', backed = 'r')
adata

AnnData object with n_obs × n_vars = 3177310 × 36406 backed at '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad'
    obs: 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intronic_reads_counted', 'donor_id', 'donor_age', 'self_reported_ethnicity_ontology_term_id', 'donor_cause_of_death', 'donor_living_at_sample_collection', 'organism_ontology_term_id', 'sample_id', 'sample_preservation_method', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'sample_collection_method', 'tissue_source', 'tissue_type', 'suspension_derivation_process', 'suspension_dissociation_reagent', 'suspension_enriched_cell_types', 'suspension_enrichment_factors', 'suspension_uuid', 'suspension_type', 'tissue_handling_interval', 'library_id', 'assay_ontology_term_id', 'sequenced_fragment', 'institute', 'library_id_repository', 'sequencing_platform', 'is_primary_data', 'cell_type_ontology_term_id', 'author_cell_type', 'disease_ontology_term_id', 'sex_ontology_term_id', 'majorclass', 'AC

In [5]:
# Count cells for each cell type in the dataset
adata.obs['cell_type'].value_counts()

cell_type
retinal rod cell                    1066056
GABAergic amacrine cell              362866
Mueller cell                         221612
OFF midget ganglion cell             200402
ON midget ganglion cell              151685
flat midget bipolar cell             144086
glycinergic amacrine cell            131207
retinal cone cell                    118559
invaginating midget bipolar cell     106152
rod bipolar cell                     101199
H1 horizontal cell                    68583
diffuse bipolar 2 cell                68176
amacrine cell                         57236
retinal bipolar neuron                52358
diffuse bipolar 1 cell                50429
diffuse bipolar 4 cell                49157
diffuse bipolar 3b cell               35342
retinal ganglion cell                 31230
giant bipolar cell                    25269
diffuse bipolar 3a cell               21827
starburst amacrine cell               20270
diffuse bipolar 6 cell                18587
OFFx cell             

## Train a scAdam model using a fraction of the dataset
The entire dataset contains `3,177,310 cells` and `36,406 genes` and occupies `35.34 GB` on disk, making it impractical to load into memory on a standard computer. Large single-cell datasets may require substantially more RAM than their file size suggests once loaded, depending on the matrix representation and associated metadata.

In addition, gene selection and model training on a dataset of this size require substantial computational time and resources.

Therefore, the scParadise team recommends extracting a smaller, representative subset of the dataset for scAdam model training.

In [6]:
# Obtain 25000 cells from dataset
adata_fraction = scparadise.scnoah.get_frac(
    path = '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad',
    fraction = 25000,
    stratify = 'cell_type', # Use to get the same ratio as in the full dataset
    random_state = 0
)

In [7]:
# Count cells for each cell type in the fraction of dataset
adata_fraction.obs['cell_type'].value_counts()

cell_type
retinal rod cell                    8388
GABAergic amacrine cell             2855
Mueller cell                        1744
OFF midget ganglion cell            1577
ON midget ganglion cell             1193
flat midget bipolar cell            1134
glycinergic amacrine cell           1032
retinal cone cell                    933
invaginating midget bipolar cell     835
rod bipolar cell                     796
H1 horizontal cell                   540
diffuse bipolar 2 cell               536
amacrine cell                        450
retinal bipolar neuron               412
diffuse bipolar 1 cell               397
diffuse bipolar 4 cell               387
diffuse bipolar 3b cell              278
retinal ganglion cell                246
giant bipolar cell                   199
diffuse bipolar 3a cell              172
starburst amacrine cell              159
diffuse bipolar 6 cell               146
OFFx cell                            122
astrocyte                            111
H2 hor

In [8]:
# Get raw counts from adata_fraction.raw
adata_fraction = adata_fraction.raw.to_adata()
# Replace variable names with gene names
adata_fraction.var.set_index('feature_name', inplace = True)
adata_fraction.var_names = adata_fraction.var_names.astype('str')
adata_fraction.var_names_make_unique()
# Normalize data
sc.pp.normalize_total(adata_fraction, target_sum = None)
sc.pp.log1p(adata_fraction)

In [9]:
# Select genes for model training (marker genes of cell types)
lst_genes = []
lst_annotations = ['majorclass', 'cell_type'] # annotation levels
for annotation in lst_annotations:
    sc.tl.rank_genes_groups(adata_fraction, 
                            groupby = annotation,
                            method = 't-test_overestim_var', pts = True)
    # Filter marker genes of cell types
    sc.tl.filter_rank_genes_groups(adata_fraction, 
                                   min_fold_change = 1.0, 
                                   min_in_group_fraction = 0.4,
                                   key_added = 'filtered_rank_genes_groups')
    # Create list of genes for model training

    for i in adata_fraction.obs[annotation].unique():
        df = sc.get.rank_genes_groups_df(adata_fraction, group = i, key = 'filtered_rank_genes_groups', pval_cutoff = 0.05)
        df['pts_comparizon'] = df['pct_nz_group']/df['pct_nz_reference']
        lst_genes.extend(df.sort_values(by = 'logfoldchanges', ascending = False).head(20)['names'].tolist())
        lst_genes.extend(df.sort_values(by = 'pts_comparizon', ascending = False).head(20)['names'].tolist())
# Remove duplicates 
lst_genes = np.unique(lst_genes).tolist()
print(f'Number of selected features: {len(lst_genes)}')

Number of selected features: 595


In [10]:
# Subset genes for model training
adata_fraction = adata_fraction[:, lst_genes]

In [11]:
# Alternative way to select genes for model training
# sc.pp.highly_variable_genes(adata_fraction,
#                             n_top_genes = 1000,
#                             subset = True)
# lst_genes = adata_fraction.var_names.tolist()

In [12]:
adata_balanced = scparadise.scnoah.balance(
    adata_fraction, 
    celltype_keys = lst_annotations
)

In [13]:
# Train scadam model using adata_fraction dataset
scparadise.scadam.train(
    adata_balanced,
    path = 'snRNAseq_human_retina', # path to save model
    model_name = 'Human_Retina_scAdam', # folder name with model
    celltype_keys = lst_annotations,
    eval_metric = ['accuracy', 'balanced_accuracy']
)

Device: cuda
Number of features: 595
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20717 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Training scAdam model:  10%|█▊                | 20/200 [01:31<13:39,  4.56s/it]

Early stopping triggered! Best score: 0.9943
Training completed!
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam


## Evaluation of model quality

For model evaluation, we use another subset of 25,000 cells generated using a different random state.

In [14]:
# Get test dataset for model quality evaluation
adata_test = scparadise.scnoah.get_frac(
    path = '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad',
    fraction = 25000,
    stratify = 'cell_type',
    random_state = 42
)

In [15]:
# Check common cells between test and training datasets
lst_train = adata_fraction.obs_names.tolist()
lst_test = adata_test.obs_names.tolist()
lst_train.extend(lst_test)
lst_train = np.unique(lst_train)
percent = round((2 * len(lst_test) - len(lst_train))/len(lst_test)*100, 5)
print(f"There are {percent} % common cells ({2 * len(lst_test) - len(lst_train)} cells) between the test and training datasets")

There are 0.836 % common cells (209 cells) between the test and training datasets


Fewer than 1% of cells overlap between the training and test datasets. Although this overlap is minimal and is unlikely to materially affect the evaluation, it should be acknowledged when interpreting the model’s performance.

In [16]:
# Apply the same preprocessing steps to the test dataset as used for training
# Get raw counts from adata_fraction.raw
adata_test = adata_test.raw.to_adata()

# Replace variable names with gene names
adata_test.var.set_index('feature_name', inplace = True)
adata_test.var_names = adata_test.var_names.astype('str')
adata_test.var_names_make_unique()

# Normalize data
sc.pp.normalize_total(adata_test, target_sum = None)
sc.pp.log1p(adata_test)
adata_test.raw = adata_test

In [17]:
# Predict cell types using trained model
adata_test = scparadise.scadam.predict(
    adata_test, 
    path_model = 'snRNAseq_human_retina/Human_Retina_scAdam'
)

Cell type annotation using the 2nd version of the scAdam model
scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Gene alignment:
  Model features: 595
  Matched features: 595 (100.0%)


Predicting: 100%|██████████████████████████████| 98/98 [00:01<00:00, 69.80it/s]


Added cell type column: pred_celltype_l1
Added probabilities column: pred_celltype_l1_probability
Added cell type column: pred_celltype_l2
Added probabilities column: pred_celltype_l2_probability


In [18]:
## Check model quality
df_l1 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype = 'majorclass',
    pred_celltype = 'pred_celltype_l1'
)
df_l1

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
AC,0.9991,0.9996,0.9998,0.9993,0.9997,0.9993,4496
Astrocyte,1.0000,0.982,1.0,0.9909,0.991,0.9802,111
BC,0.9998,0.9991,0.9999,0.9994,0.9995,0.9989,5437
Cone,1.0000,0.999,1.0,0.9995,0.9995,0.9989,1000
HC,0.9969,1.0,0.9999,0.9984,1.0,0.9999,634
MG,1.0000,0.9994,1.0,0.9997,0.9997,0.9994,1744
Microglia,1.0000,0.9487,1.0,0.9737,0.974,0.9439,39
RGC,0.9987,1.0,0.9998,0.9994,0.9999,0.9998,3144
RPE,1.0000,1.0,1.0,1.0,1.0,1.0,7
Rod,0.9995,0.9998,0.9998,0.9996,0.9998,0.9995,8388


In [19]:
## Check model quality
df_l2 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype = 'cell_type',
    pred_celltype = 'pred_celltype_l2'
)
df_l2

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
GABAergic amacrine cell,0.9916,0.9982,0.9989,0.9949,0.9986,0.9971,2855
H1 horizontal cell,0.9836,0.9981,0.9996,0.9908,0.9989,0.9976,540
H2 horizontal cell,0.9888,0.9362,1.0,0.9617,0.9675,0.9302,94
Mueller cell,1.0000,0.9994,1.0,0.9997,0.9997,0.9994,1744
OFF midget ganglion cell,0.9343,0.9112,0.9957,0.9226,0.9525,0.8996,1577
OFF parasol ganglion cell,0.9744,0.962,0.9999,0.9682,0.9808,0.9583,79
OFFx cell,1.0000,0.9754,1.0,0.9876,0.9876,0.973,122
ON midget ganglion cell,0.9411,0.9103,0.9971,0.9254,0.9527,0.8998,1193
ON parasol ganglion cell,0.9796,0.9796,1.0,0.9796,0.9897,0.9776,49
ON-blue cone bipolar cell,0.9524,0.8696,1.0,0.9091,0.9325,0.8582,23


The model performs well except for the `retinal ganglion cell`. <br> 

You could try using a different random state to generate another test dataset.

## Iterative warm start training (optional)

You may use another subset of the whole dataset to increase model generalization.

In [20]:
# Do not set the lower bound of the range to 0, which was used for the primary training of the model
for i in range(1, 4):
    # Obtain 25000 cells randomly
    adata_fraction = scparadise.scnoah.get_frac(
        path = '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad',
        fraction = 25000,
        stratify = 'cell_type',
        random_state = i
    )
    
    # Get raw counts from adata_fraction.raw
    adata_fraction = adata_fraction.raw.to_adata()
    
    # Replace variable names with gene names
    adata_fraction.var.set_index('feature_name', inplace=True)
    adata_fraction.var_names = adata_fraction.var_names.astype('str')
    adata_fraction.var_names_make_unique()
    
    # Normalize data
    sc.pp.normalize_total(adata_fraction, target_sum = None)
    sc.pp.log1p(adata_fraction)
    adata_fraction.raw = adata_fraction
    
    # Subset genes for model training
    adata_fraction = adata_fraction[:, lst_genes]
    # Balance dataset
    adata_balanced = scparadise.scnoah.balance(
        adata_fraction, 
        celltype_keys = lst_annotations
    )
    adata_balanced.raw = adata_balanced
    # Warm start requires second training dataset and path to pretrained model 
    scparadise.scadam.warm_start(
        adata_balanced,
        path = 'snRNAseq_human_retina', # path to save model
        path_model = 'snRNAseq_human_retina/Human_Retina_scAdam', # folder name with pretrained model
        model_name = 'Human_Retina_scAdam',
        celltype_keys = lst_annotations,
        eval_metric = ['accuracy', 'balanced_accuracy']
    )

scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Device: cuda
Gene alignment:
 Number of features: 595
 Matched features: 595 (100.0%)
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20717 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Warm-start model fine-tuning:  29%|███▏       | 29/100 [02:35<06:20,  5.35s/it]

Early stopping triggered! Best score: 0.9943
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam
scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Device: cuda
Gene alignment:
 Number of features: 595
 Matched features: 595 (100.0%)
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20717 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Warm-start model fine-tuning:  22%|██▍        | 22/100 [02:13<07:52,  6.05s/it]

Early stopping triggered! Best score: 0.9941
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam
scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Device: cuda
Gene alignment:
 Number of features: 595
 Matched features: 595 (100.0%)
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20717 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Warm-start model fine-tuning:  16%|█▊         | 16/100 [01:28<07:42,  5.50s/it]

Early stopping triggered! Best score: 0.9937
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam


In [21]:
# Predict cell types using trained model
adata_test = scparadise.scadam.predict(
    adata_test, 
    path_model = 'snRNAseq_human_retina/Human_Retina_scAdam'
)

Cell type annotation using the 2nd version of the scAdam model
scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Gene alignment:
  Model features: 595
  Matched features: 595 (100.0%)


Predicting: 100%|██████████████████████████████| 98/98 [00:01<00:00, 91.31it/s]


Added cell type column: pred_celltype_l1
Added probabilities column: pred_celltype_l1_probability
Added cell type column: pred_celltype_l2
Added probabilities column: pred_celltype_l2_probability


In [22]:
## Check model quality
df_warm_start_l1 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype='majorclass',
    pred_celltype='pred_celltype_l1'
)
df_warm_start_l1

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
AC,0.9982,0.9996,0.9996,0.9989,0.9996,0.9992,4496
Astrocyte,1.0000,0.982,1.0,0.9909,0.991,0.9802,111
BC,0.9993,1.0,0.9998,0.9996,0.9999,0.9998,5437
Cone,1.0000,1.0,1.0,1.0,1.0,1.0,1000
HC,1.0000,1.0,1.0,1.0,1.0,1.0,634
MG,0.9989,0.9994,0.9999,0.9991,0.9997,0.9993,1744
Microglia,1.0000,1.0,1.0,1.0,1.0,1.0,39
RGC,1.0000,0.9987,1.0,0.9994,0.9994,0.9986,3144
RPE,1.0000,1.0,1.0,1.0,1.0,1.0,7
Rod,0.9999,0.9993,0.9999,0.9996,0.9996,0.9992,8388


In [23]:
## Check model quality
df_warm_start_l2 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype='cell_type',
    pred_celltype='pred_celltype_l2'
)
df_warm_start_l2

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
GABAergic amacrine cell,0.9951,0.9975,0.9994,0.9963,0.9985,0.9967,2855
H1 horizontal cell,0.9944,0.9944,0.9999,0.9944,0.9972,0.9938,540
H2 horizontal cell,0.9684,0.9787,0.9999,0.9735,0.9892,0.9765,94
Mueller cell,0.9994,0.9994,1.0,0.9994,0.9997,0.9993,1744
OFF midget ganglion cell,0.9263,0.9569,0.9949,0.9414,0.9757,0.9484,1577
OFF parasol ganglion cell,0.9870,0.962,1.0,0.9744,0.9808,0.9583,79
OFFx cell,0.9836,0.9836,0.9999,0.9836,0.9917,0.9819,122
ON midget ganglion cell,0.9412,0.9522,0.997,0.9467,0.9744,0.9451,1193
ON parasol ganglion cell,1.0000,0.9796,1.0,0.9897,0.9897,0.9776,49
ON-blue cone bipolar cell,0.9524,0.8696,1.0,0.9091,0.9325,0.8582,23


In [24]:
pd.set_option('display.max_rows', 100)
df_l2.compare(df_warm_start_l2, keep_equal=True, align_axis = 0, result_names=('default', 'warm start'))

precision recall/sensitivity  \
GABAergic amacrine cell          default        0.9916             0.9982   
                                 warm start     0.9951             0.9975   
H1 horizontal cell               default        0.9836             0.9981   
                                 warm start     0.9944             0.9944   
H2 horizontal cell               default        0.9888             0.9362   
                                 warm start     0.9684             0.9787   
Mueller cell                     default        1.0000             0.9994   
                                 warm start     0.9994             0.9994   
OFF midget ganglion cell         default        0.9343             0.9112   
                                 warm start     0.9263             0.9569   
OFF parasol ganglion cell        default        0.9744              0.962   
                                 warm start     0.9870              0.962   
OFFx cell                        default        1.0000             0.9754   
                                 warm start     0.9836             0.9836   
ON midget ganglion cell          default        0.9411             0.9103   
                                 warm start     0.9412             0.9522   
ON parasol ganglion cell         default        0.9796             0.9796   
                                 warm start     1.0000             0.9796   
S cone cell                      default        0.9710                1.0   
                                 warm start     0.8933                1.0   
amacrine cell                    default        0.9907             0.9489   
                                 warm start     0.9844             0.9822   
diffuse bipolar 1 cell           default        0.9875             0.9975   
                                 warm start     0.9900              0.995   
diffuse bipolar 2 cell           default        0.9962             0.9869   
                                 warm start     0.9944             0.9925   
diffuse bipolar 3a cell          default        1.0000             0.9942   
                                 warm start     1.0000                1.0   
diffuse bipolar 3b cell          default        0.9964             0.9964   
                                 warm start     0.9964             0.9928   
diffuse bipolar 4 cell           default        0.9948             0.9948   
                                 warm start     0.9949                1.0   
diffuse bipolar 6 cell           default        0.9932             0.9932   
                                 warm start     0.9931             0.9795   
flat midget bipolar cell         default        0.9912             0.9974   
                                 warm start     0.9947             0.9982   
giant bipolar cell               default        0.9899             0.9899   
                                 warm start     0.9801             0.9899   
glycinergic amacrine cell        default        0.9836             0.9874   
                                 warm start     0.9874             0.9864   
invaginating midget bipolar cell default        1.0000             0.9976   
                                 warm start     0.9988                1.0   
microglial cell                  default        1.0000             0.9231   
                                 warm start     1.0000                1.0   
retinal bipolar neuron           default        0.9903             0.9951   
                                 warm start     0.9976             0.9927   
retinal cone cell                default        1.0000             0.9957   
                                 warm start     1.0000             0.9914   
retinal ganglion cell            default        0.4695              0.626   
                                 warm start     0.6444             0.4715   
retinal rod cell                 default        0.9995             0.9999   
                                 warm sta

Iterative warm start training led to an increase in all model quality metrics (rows - macro average, weighted average, accuracy, and balanced accuracy).

In [25]:
import session_info2
session_info2.session_info()

Package,Version
pandas,2.3.3
scanpy,1.11.5
scparadise,1.2.0
numpy,2.4.6
anndata,0.12.16
Component,Info
Python,"3.11.15 (main, Jun 11 2026, 15:20:16) [GCC 14.3.0]"
OS,Linux-6.8.0-138-generic-x86_64-with-glibc2.39
Updated,2026-09-04 00:04
Dependency,Version
